File: ocaml/2015/day06b/day06b.ml
open Printf
open Sys

let usage progname =
  eprintf "usage: %s <input file>\n" progname;
  exit 1

let perform grid action (x1, y1, x2, y2) =
  for j = y1 to y2 do
    for i = x1 to x2 do
      match action with
      | "turn on" -> grid.(j).(i) <- grid.(j).(i) + 1
      | "turn off" -> grid.(j).(i) <- if grid.(j).(i) > 0 then grid.(j).(i) - 1 else grid.(j).(i)
      | "toggle" -> grid.(j).(i) <- grid.(j).(i) + 2
      | _ -> ()
    done
  done

let sum grid =
  let total = ref 0 in
  for j = 0 to 999 do
    for i = 0 to 999 do
        total := !total + grid.(j).(i)
     done
  done ;
  !total

let process_file filename =
  try
    let grid = Array.init 1000 (fun _ -> Array.make 1000 0) in
    let lines = In_channel.input_lines (In_channel.open_text filename) in
    let re = Str.regexp {|\(turn on\|turn off\|toggle\) \([0-9]+\),\([0-9]+\) through \([0-9]+\),\([0-9]+\)|} in
    Seq.iter (fun line ->
      if Str.string_match re line 0 then
        let action = Str.matched_group 1 line in
        let x1 = int_of_string (Str.matched_group 2 line) in
        let y1 = int_of_string (Str.matched_group 3 line) in
        let x2 = int_of_string (Str.matched_group 4 line) in
        let y2 = int_of_string (Str.matched_group 5 line) in
        perform grid action (x1, y1, x2, y2)
    ) (List.to_seq lines) ;
    sum grid
  with
  | Sys_error msg ->
      eprintf "error reading file: %s\n" msg;
      exit 1

let () =
  let args = Sys.argv
  in if Array.length args < 2 then
       usage args.(0) ;
     let filename = args.(1) in
     let result = process_file filename in
     printf "result = %d\n" result

File: nim/2015/day06b/day06b.nim
import os, re, strutils

proc perform(grid: var array[1000*1000, int], action: string, r1: int, c1: int, r2: int, c2: int) =
  for row in r1..r2:
    for col in c1..c2:
      if action == "turn on":
        grid[row*1000 + col] += 1
      elif action == "turn off":
        grid[row*1000 + col] = max(0, grid[row*1000 + col] - 1)
      elif action == "toggle":
        grid[row*1000 + col] += 2

proc sum(grid: array[1000*1000, int]): int =
  var total = 0
  for row in 0..999:
    for col in 0..999:
      total += grid[row*1000 + col]
  total

proc processFile(filename: string): int =
  var grid: array[1000*1000, int]
  var caps: array[5, string];
  for line in lines(filename):
    if find(line, re"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)", caps) != -1:
      let action = caps[0]
      let r1 = parseInt(caps[1])
      let c1 = parseInt(caps[2])
      let r2 = parseInt(caps[3])
      let c2 = parseInt(caps[4])
      perform(grid, action, r1, c1, r2, c2)
  sum(grid)

when isMainModule:
  if paramCount() < 1:
    quit "usage: " & getAppFilename() & " <input file>", 1

  let filename = paramStr(1)
  try:
    echo "result = ", processFile(filename)
  except OSError as e:
    quit "error while processing file `" & filename & "': " & e.msg, 1
